modules/core/src/test/java/org/apache/ignite/internal/processors/cache/GridCachePartitionedClientTxMetricsTest.java

```diff
@@ -0,0 +1,216 @@
+        doPutOnClient(OPTIMISTIC, REPEATABLE_READ, true);
+    }
+
+    /**
+     *
+     */
+    public void testCommitOptimisticSerializableTxOnClient() throws IgniteInterruptedCheckedException {
+        doPutOnClient(OPTIMISTIC, SERIALIZABLE, true);
+    }
+
+    /**
+     *
+     */
+    public void testRollbackPessimisticReadCommittedTxOnClient() throws IgniteInterruptedCheckedException {
+        doPutOnClient(PESSIMISTIC, READ_COMMITTED, false);
+    }
+
+    /**
+     *
+     */
+    public void testRollbackPessimisticRepeatableReadTxOnClient() throws IgniteInterruptedCheckedException {
+        doPutOnClient(PESSIMISTIC, REPEATABLE_READ, false);
+    }
+
+    /**
+     *
+     */
+    public void testRollbackPessimisticSerializableTxOnClient() throws IgniteInterruptedCheckedException {
+        doPutOnClient(PESSIMISTIC, SERIALIZABLE, false);
+    }
+
+    /**
+     *
+     */
+    public void testRollbackOptimisticReadCommittedTxOnClient() throws IgniteInterruptedCheckedException {
+        doPutOnClient(OPTIMISTIC, READ_COMMITTED, false);
+    }
+
+    /**
+     *
+     */
+    public void testRollbackOptimisticRepeatableReadTxOnClient() throws IgniteInterruptedCheckedException {
+        doPutOnClient(OPTIMISTIC, REPEATABLE_READ, false);
+    }
+
+    /**
+     *
+     */
+    public void testRollbackOptimisticSerializableTxOnClient() throws IgniteInterruptedCheckedException {
+        doPutOnClient(OPTIMISTIC, SERIALIZABLE, false);
+    }
+
+    /**
+     * @param transactionConcurrency Transaction concurrency.
+     * @param transactionIsolation Transaction isolation.
+     * @param commit {@code True} if should commit, otherwise - rollback transaction.
+     */
+    private void doPutOnClient(TransactionConcurrency transactionConcurrency,
+        TransactionIsolation transactionIsolation, boolean commit) throws IgniteInterruptedCheckedException {
+        Ignite client = ignite(gridCount() - 1);
+
+        IgniteCache<Integer, String> cache = client.cache(DEFAULT_CACHE_NAME);
+
+        IgniteTransactions transactions = client.transactions();
+
+        for (int i = 0; i < EXPECTED_TXS_FINISH_COUNT; i++) {
+            try (Transaction tx = transactions.txStart(transactionConcurrency, transactionIsolation)) {
+                cache.put(i, "value");
+
+                if (commit)
+                    tx.commit();
+                else
+                    tx.rollback();
+            }
+        }
+
+        if (commit) {
+            GridTestUtils.waitForCondition(() -> {
+                boolean metricsUpdated =
+                    client.cache(DEFAULT_CACHE_NAME).metrics().getCacheTxCommits() == EXPECTED_TXS_FINISH_COUNT
+                        && transactions.metrics().txCommits() == EXPECTED_TXS_FINISH_COUNT;
+
+                return metricsUpdated;
+
+            }, 3 * METRICS_UPDATE_FREQUENCY);
+        }
+        else {
+            GridTestUtils.waitForCondition(() -> {
```

Comment:
We should check that waitForCondition() returns true, because in case of timeout if will return false - that means we didn't recieve expected metrics.